proposals/47-js-trim-and-modular.md

```diff
@@ -0,0 +1,124 @@
+
+1. There are a _lot_ of dependencies
+   1. Takes a long time to install
+   2. Assumes the user wants _all the things_
+2. The total weight of code is very large
+   1. Related to the _all the things_ accumulative design of the stack
+   2. Crypto libraries mostly to blame
+3. The API surface area is very large
+   1. Related to the projection of the CLI API up into the browser API and exposing _all the things_
+   2. The complexity of the entire stack means there's a lot of different ways to achieve things using different APIs and configuring them separately, but you still get _all the things_
+
+The goal of this project is to initial exploration, design and planning of a suite of JavaScript libraries that can be composed according to user / application needs, with minimal inclusion of features and functionality that are out of bounds for those needs.
+
+Actual development within this project would be exploratory / experimental only. Outcomes would be further scoped proposals for more work to do actual implementation of parts, or all of the proposed suite.
+
+#### Assumptions &amp; hypotheses
+
+_What must be true for this project to matter?_
+
+ * Interacting with IPFS in JavaScript is something that users want to do
+ * Current js-ipfs is too monolithic and causing barriers to entry
+
+#### User workflow example
+
+_How would a developer or user use this new capability?_
+
+Developers would be able to choose the level of interaction with our web3 stack in JavaScript:
+
+ * IPLD ("I just want to make CIDs for DAG-CBOR thanks")
+ * libp2p ("I just want to chat bitswap with known peers and handle everything else myself thanks")
+ * UnixFS ("I want to wrap up this in-browser video recording into DAG-PB in a CAR file and let the user download it, thanks")
+ * CIDs ("I want to make this thing into an NFT and would like to know the CID before I submit it to `<IPFS, Filecoin, service, etc.>`")
+ * etc.
+
+Some of these kinds of use-cases are either impossible or very difficult with the current incarnation of the JavaScript stack.
+
+#### Impact
+
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+Reach JavaScript developers at their level of comfort. 
+
+Surveys like [the one conducted annually by StackOverflow](https://insights.stackoverflow.com/survey/2019/#most-popular-technologies), suggest that there are nearly an order of magnitude more developers who know Javascript (67%) than who know Go (8%). But our JavaScript offering is not friendly for general consumption and we usually end up pushing them to use go-ipfs (and friends) and build wrappers around it. Many use-cases are much simpler than requiring the entire js-ipfs stack as it currently exists and we're not serving those (well).
+
+#### Leverage
+
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+#### Confidence
```

Comment:
As a data point to increase confidence here, Textile did not use our libs, but built their own ipld+libp2p variant.

@Gozala was looking into creating something like this a while ago and ended up transfering the project from Textile to our org. 
It now lives at https://github.com/ipfs/js-dag-service

Reply:
There was some great work there, this proposal would definitely take those learnings on board.